File: app/src/main/java/com/moh/mhealth/Global_Helper.java
package com.moh.mhealth;

import android.content.Intent;
import android.util.Log;
import android.widget.EditText;

import com.moh.mhealth.pages.Diagnostics1;
import com.moh.mhealth.pages.Diagnostics2;
import com.moh.mhealth.pages.Diagnostics3;
import com.moh.mhealth.pages.Diagnostics4;
import com.moh.mhealth.pages.NewPatient;
import com.moh.mhealth.pages.PatientList;
import com.moh.mhealth.pages.PatientSurvey;

import java.time.LocalDate;
import java.util.Date;

public class Global_Helper {

    public static final String USERNAME = "user"; // Gonna need to fix
    public static final String PASSWORD = "1234"; // Gonna need to fix

    private static final Class[] diagOrder = {
            PatientList.class,
            NewPatient.class,
            PatientSurvey.class,
            Diagnostics1.class,
            Diagnostics2.class,
            Diagnostics3.class,
            Diagnostics4.class,
    };

    private static Patient currentPatient;

    private static int currentDiag = 0;

    private static boolean isMetric;


    public static double getLbToKg(double lb) {
        return Math.round(lb * 0.45359237 * 100.0) / 100.0;
    }
    public static double getInchToCm(double in) {
        return Math.round(in * 2.54 * 100.0) / 100.0;
    }
    public static double getFToC(double f) {
        return Math.round((f - 32) * (5/9.0) * 100.0) / 100.0;
    }

    public static double getKgToLb(double kg) {
        return Math.round(kg / 0.45359237 * 100.0) / 100.0;
    }
    public static double getCmToInch(double cm) {
        return Math.round(cm / 2.54 * 100.0) / 100.0;
    }
    public static double getCToF(double c) {
        return Math.round(((9/5.0 * c) + 32) * 100.0) / 100.0;
    }

    public static double getDoubleFromEditText(EditText editText) {
        String doubleStr = editText.getText().toString();

        if (!doubleStr.isEmpty()) {
            try {
                return Double.parseDouble(doubleStr);
            } catch (Exception e1) {
                // I don't really care about the error I don't think...
            }
        }
        return -1;
    }

    public static int getIntFromEditText(EditText editText) {
        String intStr = editText.getText().toString();

        if (!intStr.isEmpty()) {
            try {
                return Integer.parseInt(intStr);
            } catch (Exception e1) {
                // I don't really care about the error I don't think...
            }
        }
        return -1;
    }

    public static void createNewPatient(String name, char gender, LocalDate dob) {
        currentPatient = new Patient(name, gender, dob);
    }


    public static void addPatientSurvey(String village, int[] distanceTraveled, int[] timeWaited) {
        currentPatient.setVillage(village);
        currentPatient.setDistance_traveled(distanceTraveled);
        currentPatient.setTime_waited(timeWaited);
    }
    public static void setUnits(int unitsInt) {
        isMetric = (unitsInt == 0) ? true : false;
    }

    public static void addDiag1Metric(double kg, int cm, double tempC) {
        currentPatient.setWeight(kg);
        currentPatient.setHeight(cm);
        currentPatient.setTemp(tempC);
    }
    // Wrapper to both save diag1 variables and convert imperial to metric
    public static void addDiag1Imperial(double lbs, int ft, int inch, double tempF) {
        addDiag1Metric(getLbToKg(lbs), (int) getInchToCm(ft * 12 + inch), getFToC(tempF));
    }

    public static void addDiag2(int[] bp, int p, int rr, double muacCM) {
        currentPatient.setBp(bp);
        currentPatient.setP(p);
        currentPatient.setRr(rr);
        currentPatient.setMuac(muacCM);
    }

    public static void addDiag3(String pres, String assess) {
        currentPatient.setPres(pres);
        currentPatient.setAssess(assess);
    }

    public static void addDiag4(String diag, String treat) {
        currentPatient.setDiag(diag);
        currentPatient.setTreat(treat);
    }

    public static boolean isMetric() {
        return isMetric;
    }

    public static Patient getCurrentPatient() {
        return currentPatient;
    }

    public static Class nextDiag() {
        Log.i("NextDiag", currentDiag + ":: " + diagOrder[currentDiag].getName());
        assert (currentDiag < diagOrder.length - 1);
        currentDiag++;
        return diagOrder[currentDiag];
    }

    public static Class prevDiag() {
        assert (currentDiag > 0);

        if (currentDiag == 1) { // Gonna be returning to the patient list page, so cancel current patient
            currentPatient = null;
        }
        currentDiag--;
        return diagOrder[currentDiag];
    }

    public static Class endPatient() {
        currentDiag = 0;
        currentPatient = null;

        isMetric = true;

        return diagOrder[0];
    }
}

File: app/src/main/java/com/moh/mhealth/GlobalHelper.kt
package com.moh.mhealth

import android.util.Log
import android.widget.EditText
import com.moh.mhealth.pages.*
import com.moh.mhealth.objects.Patient
import java.time.LocalDate
import kotlin.math.roundToInt

object GlobalHelper {
    const val USERNAME = "" // Gonna need to fix
    const val PASSWORD = "" // Gonna need to fix
    private val diagOrder = arrayOf<Class<*>>(
        PatientList::class.java,
        NewPatient::class.java,
        PatientSurvey::class.java,
        Diagnostics1::class.java,
        Diagnostics2::class.java,
        Diagnostics3::class.java,
        Diagnostics4::class.java
    )
    var currentPatient: Patient? = null
        private set
    private var currentDiag = 0
    var isMetric = false

    fun getLbToKg(lb: Double): Double {
        return (lb * 0.45359237 * 100.0).roundToInt() / 100.0
    }

    fun getInchToCm(inch: Double): Double {
        return (inch * 2.54 * 100.0).roundToInt() / 100.0
    }

    fun getFToC(f: Double): Double {
        return ((f - 32) * (5 / 9.0) * 100.0).roundToInt() / 100.0
    }

    fun getKgToLb(kg: Double): Double {
        return (kg / 0.45359237 * 100.0).roundToInt() / 100.0
    }

    fun getCmToInch(cm: Double): Double {
        return (cm / 2.54 * 100.0).roundToInt() / 100.0
    }

    fun getCToF(c: Double): Double {
        return ((9 / 5.0 * c + 32) * 100.0).roundToInt() / 100.0
    }

    fun getDoubleFromEditText(editText: EditText?): Double {
        val doubleStr = editText!!.text.toString()
        if (doubleStr.isNotEmpty()) {
            try {
                return doubleStr.toDouble()
            } catch (e1: Exception) {
                // I don't really care about the error I don't think...
            }
        }
        return -1.0
    }

    fun getIntFromEditText(editText: EditText?): Int {
        val intStr = editText!!.text.toString()
        if (intStr.isNotEmpty()) {
            try {
                return intStr.toInt()
            } catch (e1: Exception) {
                // I don't really care about the error I don't think...
            }
        }
        return -1
    }

    fun createNewPatient(name: String, gender: Char, dob: LocalDate?) {
        currentPatient = Patient(name, gender, dob)
    }

    fun addPatientSurvey(village: String, distanceTraveled: IntArray?, timeWaited: IntArray?) {
        currentPatient?.village = village
        currentPatient?.distanceTraveled = distanceTraveled
        currentPatient?.timeWaited = timeWaited
    }

    fun addDiag1Metric(kg: Double, cm: Int, tempC: Double) {
        currentPatient?.weight = kg
        currentPatient?.height = cm
        currentPatient?.temp = tempC
    }

    // Wrapper to both save diag1 variables and convert imperial to metric
    fun addDiag1Imperial(lbs: Double, ft: Int, inch: Int, tempF: Double) {
        addDiag1Metric(
            getLbToKg(lbs),
            getInchToCm((ft * 12 + inch).toDouble()).toInt(),
            getFToC(tempF)
        )
    }

    fun addDiag2(bp: IntArray?, p: Int, rr: Int, muacCM: Double) {
        currentPatient?.bp = bp
        currentPatient?.p = p
        currentPatient?.rr = rr
        currentPatient?.muac = muacCM
    }

    fun addDiag3(pres: String, assess: String) {
        currentPatient?.pres = pres
        currentPatient?.assess = assess
    }

    fun addDiag4(diag: String, treat: String) {
        currentPatient?.diag = diag
        currentPatient?.treat = treat
    }

    fun nextDiag(): Class<*> {
        Log.i("NextDiag", currentDiag.toString() + ":: " + diagOrder[currentDiag].name)
        assert(currentDiag < diagOrder.size - 1)
        currentDiag++
        return diagOrder[currentDiag]
    }

    fun prevDiag(): Class<*> {
        assert(currentDiag > 0)
        if (currentDiag == 1) { // Gonna be returning to the patient list page, so cancel current patient
            currentPatient = null
        }
        currentDiag--
        return diagOrder[currentDiag]
    }

    fun endPatient(): Class<*> {
        currentDiag = 0
        currentPatient = null
        isMetric = true
        return diagOrder[0]
    }
}